java/android-platform-tools/Portfile

```diff
@@ -29,6 +29,37 @@ use_configure       no
 
 worksrcdir          platform-tools
 
+# detect which version to serve to Intel clients based on HW capabilities
+platform i386 {
+    if {[catch {sysctl hw.optional.x86_64} x86_64] || $x86_64 == 1} {
+        if {[catch {sysctl hw.optional.sse4_1} sse4] || $sse4 == 0} {
+            default_variants    +no_sse4
```

Comment:
I took this approach because, at least to me, separating the detection logic from the variant definitions seems clean from a Portfile intelligibility standpoint (and easier to update again when, sometime in the future, Google starts shipping adb as an arm64 binary only). Plus, from my testing, it correctly selects the right variant based on the user's CPU, and lets them install the other legacy variants manually if for some reason they want to do so.

My experience is very limited, so if my approach is deemed unacceptable I'll try to figure out something else.

Reply:
I can't experiment with this right now (away from any Mac), but the potential problem I see is that Portfiles are parsed twice. First to get the lost of variants them again to evaluate it with the variant selection.

By changing the option `version` between these two parses could have an effect. However, I am not sure how it behaves.

Usually we have conditionals such as `if {${os.platform} == ...}` to make these distinctions for legacy OS versions. What is the advantage of using variants for this?

Or is there a use case to explicitly request +i386 on let's say arm64? But with that you would get a binary archive for arm64 containing i386 binaries. I think that will potentially even trigger on rev-upgrade?

Reply:
I did not use `${os.platform}` because the legacy variants I'm proposing aren't really linked to specific OS versions, they're specifically linked to CPU capabilities.

I found this issue after [updating **scrcpy** to run on older OS versions](https://github.com/macports/macports-ports/pull/20727), as the current version of **adb** would not run on my MBP with a Merom Core 2 Duo CPU, which lacks SSE4 support (introduced in the subsequent Penryn revision). Which in turn led me to the trac ticket I mentioned in this PR.
This Mac is running El Capitan, but distinguishing by OS would not really work, as El Capitan can be run on much later machines, with SSE4 support, which do not need the older version.
The same can be said about the 32-bit i386 variant: 32-bit Intel Macs support 10.6 at a maximum, but 10.6 can be run on x86_64 Macs too.

I'm not familiar with MacPort internals, so I had no idea of the potential pitfalls you mention; all I can say is I did some testing on my two Intel Macs (one without, one with SSE4 support), and it consistently chose the right variant in my tests. (I've also tried inverting some conditions to make it select a different variant on purpose, and it always selected the intended one).

As for why someone would want to request a legacy version even if it's not needed on their specific machine, the one that comes to my mind would be future testing for new versions of **scrcpy**, to see if they still work with the old **adb**.
I don't have any arm64 Macs, so I can't offer any testing on that front.

Once again, of course the best thing would be - assuming it is possible - building tailored variants of the current version, instead of serving binaries of old versions, but that's way out of my reach, so this workaround is the only way I can contribute.

Reply:
@jmroot Can you take a look whether using variants for a platform-specific `version` is as bad as I think it could be?

Reply:
I consider changing a port's version dynamically to always be a problem. This is because some operations (for various reasons, often performance) use information from the PortIndex, which is static. Changing based on `os.major` is bad enough, but indeed changing based on the runtime CPU arch is a bit worse, since you get the same PortIndex regardless of whether you're on i386 or x86_64.

The solution I always recommend is to provide the older versions as separate ports (not necessarily separate Portfiles, they can be subports). In the case that an older version is needed, the main port installs nothing (in practice just a Readme file) and depends on the appropriate older version port. Dynamically changing dependencies are handled well.

BTW, the check for whether an older version is needed should start by looking at `configure.build_arch`, not the CPU reported by sysctl. In the case that it's x86_64, the check for SSE4 will still need to use sysctl.

Reply:
Wouldn't that require duplicating the detection logic in `scrcpy`, so it can specify different dependences based on the user's CPU?

Reply:
If `scrcpy` just depends on `android-platform-tools`, then it will get the correct version, because the idea is that `android-platform-tools` depends on the appropriate version-specific port for the CPU.